tools/mqtt-bench/client.go

```diff
@@ -0,0 +1,227 @@
+package main
+
+import (
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/GaryBoone/GoStats/stats"
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+)
+
+type SubTimes map[string][]float64
+
+// Client - represents mqtt client
+type Client struct {
+	ID         string
+	BrokerURL  string
+	BrokerUser string
+	BrokerPass string
+	MsgTopic   string
+	MsgSize    int
+	MsgCount   int
+	MsgQoS     byte
+	Quiet      bool
+	mqttClient *mqtt.Client
+	Mtls       bool
+	SkipTlsVer bool
+	CA         []byte
+	clientCert tls.Certificate
+	ClientKey  *rsa.PrivateKey
+}
+
+// RunPublisher - runs publisher
+func (c *Client) RunPublisher(res chan *RunResults, mtls bool) {
+	newMsgs := make(chan *Message)
+	pubMsgs := make(chan *Message)
+	doneGen := make(chan bool)
+	donePub := make(chan bool)
+	runResults := new(RunResults)
+
+	started := time.Now()
+	// start generator
+	go c.genMessages(newMsgs, doneGen)
+	// start publisher
+	go c.pubMessages(newMsgs, pubMsgs, doneGen, donePub, mtls)
+
+	times := []float64{}
+	for {
+		select {
+		case m := <-pubMsgs:
+			cid := m.ID
+			if m.Error {
+				////log.Printf("CLIENT %v ERROR publishing message: %v: at %v\n", c.ID, m.Topic, m.Sent.Unix())
+				runResults.Failures++
+			} else {
+				//log.Printf("Message published: %v: sent: %v delivered: %v flight time: %v\n", m.Topic, m.Sent, m.Delivered, m.Delivered.Sub(m.Sent))
+				runResults.Successes++
+				runResults.ID = cid
+				times = append(times, float64(m.Delivered.Sub(m.Sent).Nanoseconds()/1000)) // in microseconds
+			}
+		case <-donePub:
+			// calculate results
+			duration := time.Now().Sub(started)
+			runResults.MsgTimeMin = stats.StatsMin(times)
+			runResults.MsgTimeMax = stats.StatsMax(times)
+			runResults.MsgTimeMean = stats.StatsMean(times)
+			runResults.MsgTimeStd = stats.StatsSampleStandardDeviation(times)
+			runResults.RunTime = duration.Seconds()
+			runResults.MsgsPerSec = float64(runResults.Successes) / duration.Seconds()
+
+			// report results and exit
+			res <- runResults
+			return
+		}
+	}
+}
+
+// RunSubscriber - runs a subscriber
+func (c *Client) RunSubscriber(wg *sync.WaitGroup, subTimes *SubTimes, done *chan bool, mtls bool) {
+	defer wg.Done()
+	// start generator
```

Comment:
Confusing comments - do we start generator or subscriber?

Reply:
generator of messages , function below